### ISSUE TITLE
How to determine the disableApiTermination attribute value 

### ISSUE THREAD

**Issue by nilroy** [fog/fog-aws#98]
How do I determine whether API termination is disbaled or enabled using fog. The disable_api_termination atribute of the fog instance object seems to be nil always.

**Comment by nilroy** [comment]
Adding @geemus

**Comment by geemus** [comment]
My guess is that, although we added support for it when creating/modifying servers, it never made it's way into the parser code for describe instances. Searching the parser seems to confirm that, see: https://github.com/fog/fog-aws/blob/master/lib/fog/aws/parsers/compute/describe_instances.rb

Would you be up for adding that? It will just need to be added in the right context (which you should be able to figure out from AWS docs, but I'm happy to help if you have questions or concerns). Thanks!

**Comment by nilroy** [comment]
Thanks for your help...I shall go through the doc and see.

**Comment by nilroy** [comment]
I am afraid that AWS supports it from describe instances. http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/ApiReference-cmd-DescribeInstances.html. But AWS does support describe-instance-attribute 

[system] [~] » aws ec2 describe-instance-attribute --instance-id i-xxxxx --attribute "disableApiTermination"
------------------------------
|  DescribeInstanceAttribute |
+-------------+--------------+
|  InstanceId |  i-xxxx |
+-------------+--------------+
||   DisableApiTermination  ||
|+-------------+------------+|
||  Value      |  True      ||
|+-------------+------------+|

Is there a fog way?Or that needs to be implemented?

**Comment by geemus** [comment]
Hmm, yeah. I guess maybe just using that method would be best. Perhaps we could have an accessor that would lazily load it (so if you called that method and it didn't already have a value, it could do this call). What do you think?

**Comment by nilroy** [comment]
@geemus that sounds great..

**Comment by geemus** [comment]
Would you be up for attempting a pull request to implement this functionality? Thanks!

**Comment by nilroy** [comment]
I shall attempt it. I am new to this project so might take a bit longer. By the way just to be double sure by accessor you meant a method?

**Comment by nilroy** [comment]
@geemus. I came across one AWS API reference http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html.  And I also found the "request" method (https://github.com/fog/fog-aws/blob/5903a7819b12ec6feb3b1d4177f9a1f4a4ca3a59/lib/fog/aws/federation.rb) method. Now the DescribeInstance action does not produce the disable_api_termination attribute in response. 
I think you asked to use DescribeInstanceAttribute action (http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceAttribute.html). I think I did not understand your previous comment fully. Sorry if I am confusing.

**Comment by geemus** [comment]
Hmm. I guess this might be slightly more complicated than I realized.

There would be a few steps to this I think. First, you would need to add a describe_instance_attribute request (I didn't realize this was not-yet implemented). It would use request, similar to the example from describe_instances you included. That "request" method uses excon and should automatically setup the proper endpoint and stuff (you just need to specify the unique parts of the request, as it defaults to the more common settings).

You'll also need to add a new parser to actually turn the XML into a ruby data structure, you can see an example of this at: https://github.com/fog/fog-aws/blob/master/lib/fog/aws/parsers/compute/describe_instances.rb (this is one of the harder/more annoying parts unfortunately).

Once that is done, within the model I was thinking something (roughly) like:
```
def disable_api_termination
  @disable_api_termination ||= begin
    connection.describe_instance_termination['DisableApiTermination']
  end
end
```

Does that make sense/help? Definitely let me know if you still have questions or concerns. Thanks!

**Comment by nilroy** [comment]
@geemus  Thanks a lot for your comment. I shall give a try to implement it and comeback if I face some more issue.

**Comment by geemus** [comment]
Sounds great, thanks!

**Comment by nilroy** [comment]
@geemus  : One small question. Is there a specific version of AWS API to use? The current version is 2015-03-01. I am planning to use this version

**Comment by nilroy** [comment]
@geemus : Maybe describe instance attribute request does not support signature v4. I could get describe instances action output with the script.I shall try with signature v2. But anyways I think rest of the outputs would have similar XML structure. I would try to make the parser for few attributes and then add more.

**Comment by nilroy** [comment]
@geemus :- I am now able to make the AWS API calls.. I just changed the request parameters like following

request_parameters = 'Action=DescribeInstanceAttribute&Attribute=instanceType&InstanceId=i-xxxxx&Version=2015-03-01'

Its very strange that only by swapping the position of the Attribute and InstanceId did the trick.. Anyways I shall try implementing all the attributes in the parser :)

**Comment by nilroy** [comment]
@geemus : I need one input regarding the parser. 
The Output from AWS API for block device mapping attribute looks like this

<DescribeInstanceAttributeResponse xmlns="http://ec2.amazonaws.com/doc/2015-03-01/">
    <requestId>40b06184-67f6-4dc2-b8ec-530c4a152faa</requestId>
    <instanceId>i-xxxx</instanceId>
    <blockDeviceMapping>
        <item>
            <deviceName>/dev/sda1</deviceName>
            <ebs>
                <volumeId>vol-xxxx</volumeId>
                <status>attached</status>
                <attachTime>YYYY-MM-DD hh:mm:ss.000Z</attachTime>
                <deleteOnTermination>true</deleteOnTermination>
            </ebs>
        </item>
        <item>
            <deviceName>/dev/sdf</deviceName>
            <ebs>
                <volumeId>vol-yyyy</volumeId>
                <status>attached</status>
                <attachTime>YYYY-MM-DD hh:mm:ss.000Z</attachTime>
                <deleteOnTermination>false</deleteOnTermination>
            </ebs>
        </item>
    </blockDeviceMapping>
</DescribeInstanceAttributeResponse>
I am following the aws API and trying to output as close to the AWS API does. By doing that the blockdeviceMapping attribute output is as follows:
=> #<Excon::Response:0x007fd1c13719b0
 @body="",
 @data=
  {:body=>
    {"requestId"=>"aaaa-vvvv-dddd-eeee",
     "instanceId"=>"i-xxxxx",
     "blockDeviceMapping"=>
      [{"ebs"=>{"volumeId"=>"vol-xxxxx", "status"=>"attached", "attachTime"=>YYYY-MM-DD hh:mm:ss UTC, "deleteOnTermination"=>true},
        "deviceName"=>"/dev/sda1"},
       {"ebs"=>{"volumeId"=>"vol-yyyy", "status"=>"attached", "attachTime"=>YYYY-MM-DD hh:mm:ss UTC, "deleteOnTermination"=>false},
        "deviceName"=>"/dev/sdf"}]},

But I have noticed that in describe_instance parser the block device mapping is done a bit different way

{"blockDeviceMapping"=>
            [{"deviceName"=>"/dev/sda1",
              "volumeId"=>"vol-xxxxx",
              "status"=>"attached",
              "attachTime"=>YYYY-MM-DD hh:mm:ss UTC,
              "deleteOnTermination"=>true},
             {"deviceName"=>"/dev/sdf",
              "volumeId"=>"vol-yyyy",
              "status"=>"attached",
              "attachTime"=>YYYY-MM-DD hh:mm:ss UTC,
              "deleteOnTermination"=>false}]}

What do you say? Shall I do it so that it resembles the current output of describe_instances request or I keep staying close to the AWS API output.

**Comment by nilroy** [comment]
@geemus I have written the request and the parser. I am not sure how to create the model for the that. Could you please help me out?

**Comment by nilroy** [comment]
@geemus I decided to have the block_device_mappings output in describe_instace_attribute (request)  value to resemble the same as that of describe_instance(request). But I need your input in models.

**Comment by nilroy** [comment]
I have created a new pull request

**Comment by geemus** [comment]
@nilroy cool, thanks. I'll close this for now as we work through the pull request. We can definitely come back to figure out model stuff later. Thanks!
